R/getTimeDivergence.R

```diff
@@ -33,73 +33,92 @@
-#'@export
-getTimeDivergence <- function(se, field, time_field, time_interval){
-
-# Define the names for the newly added fields
-# for the time difference and community dissimilarity
-new_field <- "time_divergence"
-new_field2 <- "time_difference"
-
-# Dissimilarity will be calculated between rows, as in stats::dist
+#' @rdname getTimeDivergence
+#'
+#' @export
 check_pairwise_dist <- function (x,
                                  distfun,
                                  time_interval,
-                                 new_field,
-                                 new_field2,
-                                 time_field){
+                                 new_field = "time_divergence",
+                                 new_field2 = "time_difference",
+                                 time_field,
+                                 abund_values){
 
-    mat <- t(assay(x))
+  mat <- t(assay(x, abund_values = "counts"))
 
-    time <- colData(x)[, time_field]
+  time <- colData(x)[, time_field]
 
-    # Add new field to coldata
-    colData(x)[, new_field] <- rep(NA, nrow(mat))
-    colData(x)[, new_field2] <- rep(NA, nrow(mat))
+  ## Add new field to coldata
+  colData(x)[, new_field] <- rep(NA, nrow(mat))
+  colData(x)[, new_field2] <- rep(NA, nrow(mat))
 
+  if (nrow(mat) > time_interval) {
 
-    if (nrow(mat) > time_interval) {
+    ##beta diversity calculation
+    n <- sapply((time_interval+1):nrow(mat),
+                function (i) {distfun(mat[c(i, i-time_interval), ])})
 
-        #beta diversity calculation
-        n <- sapply((time_interval+1):nrow(mat),
-                    function (i) {distfun(mat[c(i, i-time_interval), ])})
+    for(i in (time_interval+1):nrow(colData(x))){
+      colData(x)[, new_field][[i]] <- n[[i-time_interval]]
+    }
 
-        for(i in (time_interval+1):nrow(colData(x))){
-            colData(x)[, new_field][[i]] <- n[[i-time_interval]]
-        }
+    ##time difference calculation
+    time <- sapply((time_interval+1):nrow(mat),
+                   function (i) {diff(colData(x)[c(i-time_interval, i), time_field])})
 
-        #time difference calculation
-        time <- sapply((time_interval+1):nrow(mat),
-                       function (i) {diff(colData(x)[c(i-time_interval, i), time_field])})
+    for(i in (time_interval+1):nrow(colData(x))){
+      colData(x)[, new_field2][[i]] <- time[[i-time_interval]]
 
-        for(i in (time_interval+1):nrow(colData(x))){
-            colData(x)[, new_field2][[i]] <- time[[i-time_interval]]
-        }
     }
-
     return(x)
+  }
 }
 
+#' @rdname getTimeDivergence
+#' @name getTimeDivergence
+#' @export
+getTimeDivergence <- function(se, field, time_field, time_interval, new_field = "time_divergence",
+                              new_field2 = "time_difference", abund_values = "counts" ){
+
 # Split SE into a list, by subject
 spl <- split(colnames(se), colData(se)[, field])
 
+spl_more <- spl[lapply(spl,length)>1]
+
+spl_one <- spl[lapply(spl,length) == 1]
+
```

Comment:
I tried to make the function faster and came up with this. Here are the individuals that have more than one sample separated and instead of all of them, only they are applied to `check_pairwise_dist` .  The ones that have only one sample goes through the `se[, spl_one[[i]]]` later on. Eventually these lists are put together and combined as SE object. Hitchip1006 would run in 180 seconds for the entire dataset, but the old version was running it for 250 seconds. Actually, it didn't save time as much as I hoped, so we can go back to the previous version if you think that it was more robust/efficient.

Reply:
Ok, just let's try to get something now that does the job and works as expected (unit tests). Optimization can be done later.

Anyway take a subset of hitchip for examples, examples should always be very fast.